topics/go/algorithms/data/hash/hash.go

```diff
@@ -0,0 +1,107 @@
+// Package hash implements a hash table
+package hash
+
+import (
+	"fmt"
+	"hash/maphash"
+)
+
+const (
+	numBuckets = 256 // Number of buckets in hash
+)
+
+// entry in the hash table
+type entry struct {
+	key   string
+	value int
+}
+
+// Hash is a simple Hash table
+type Hash struct {
+	buckets [][]*entry
+	h       maphash.Hash
+}
+
+// NewHash returns a new hash table
+func NewHash() *Hash {
+	return &Hash{
+		buckets: make([][]*entry, numBuckets),
+	}
+}
+
+// Set sets the key to value in the hash
+func (h *Hash) Set(key string, value int) {
+	b := h.hashKey(key)
+	// Look for existing key
+	for _, e := range h.buckets[b] {
+		if e.key == key {
+			e.value = value
+			return
+		}
+	}
+
+	h.buckets[b] = append(h.buckets[b], &entry{key, value})
+}
+
+// Get gets the value associated with key, return a error is key not found
+func (h *Hash) Get(key string) (int, error) {
+	b := h.hashKey(key)
+	for _, e := range h.buckets[b] {
+		if e.key == key {
+			return e.value, nil
+		}
+	}
+
+	return 0, fmt.Errorf("%q not found", key)
+}
+
+// Delete deletes an entry from the hash, return an error if not found
+func (h *Hash) Delete(key string) error {
+	b := h.hashKey(key)
+	bucket := h.buckets[b]
+	for i, e := range bucket {
+		if e.key == key {
+			copy(bucket[i:], bucket[i+1:])
+			bucket = bucket[:len(bucket)-1]
+			// Free memory when bucket shrinks a lot
+			if cap(bucket) > 2*len(bucket) {
+				nb := make([]*entry, len(bucket))
+				copy(nb, bucket)
+				bucket = nb
+			}
+			h.buckets[b] = bucket
+			return nil
+		}
+	}
+
+	return fmt.Errorf("%q not found", key)
+}
+
+// Len return the number of elements in the hash
+func (h *Hash) Len() int {
```

Comment:
This seems expensive. Do real implementations keep this meta-data?

Reply:
Maybe we don't want to teach keeping meta-data?

Reply:
This is expensive, and real implementations probably keep length metadata. However it's a bit tricky to keep it updated. In `Delete` it obvious but in `Set` it's only if it's a new key. I prefer the simpler implementation.

Reply:
Ok, maybe we leave a comment in the function header?

Reply:
Added a TODO comment.